libraries/Adapters/Microsoft.Bot.Builder.Adapters.Twilio/TwilioAdapter.cs

```diff
@@ -0,0 +1,248 @@
+        /// </summary>
+        /// <param name="request">A request object from Restify or Express.</param>
+        /// <param name="response">A response object from Restify or Express.</param>
+        /// <param name="bot">A bot with logic function in the form `async(context) => { ... }`.</param>
+        /// <param name="cancellationToken">A cancellation token for the task.</param>
+        /// <returns>A <see cref="Task"/> representing the asynchronous operation.</returns>
+        public async Task ProcessAsync(HttpRequest request, HttpResponse response, IBot bot, CancellationToken cancellationToken = default)
+        {
+            response.StatusCode = 200;
+            await response.WriteAsync(string.Empty, cancellationToken).ConfigureAwait(false);
+
+            var twilioSignature = request.Headers["x-twilio-signature"];
+
+            var validationUrl = _options.ValidationUrl ?? (request.Headers["x-forwarded-proto"][0] ?? request.Protocol + "://" + request.Host + request.Path);
+
+            var requestValidator = new RequestValidator(_options.AuthToken);
+
+            using (var bodyStream = new StreamReader(request.Body))
+            {
+                var payload = bodyStream.ReadToEnd();
+
+                var formattedPayload = payload.Replace("+", "%20");
+
+                var body = QueryStringToDictionary(formattedPayload);
+
+                if (!string.IsNullOrWhiteSpace(twilioSignature) &&
+                    requestValidator.Validate(validationUrl, body, twilioSignature))
+                {
+                    var jsonBody = JsonConvert.SerializeObject(body);
+
+                    var twilioEvent = JsonConvert.DeserializeObject<TwilioEvent>(jsonBody);
+
+                    if (int.TryParse(twilioEvent.NumMedia, out var numMediaResult) && numMediaResult > 0)
+                    {
+                        // specify a different event type
+                        twilioEvent.EventType = "picture_message";
+                    }
+
+                    var activity = new Activity()
+                    {
+                        Id = twilioEvent.MessageSid,
+                        Timestamp = new DateTime(),
+                        ChannelId = "twilio-sms",
+                        Conversation = new ConversationAccount() { Id = twilioEvent.From, },
+                        From = new ChannelAccount() { Id = twilioEvent.From, },
+                        Recipient = new ChannelAccount() { Id = twilioEvent.To, },
+                        Text = twilioEvent.Body,
+                        ChannelData = twilioEvent,
+                        Type = ActivityTypes.Message,
+                    };
+
+                    // create a conversation reference
+                    using (var context = new TurnContext(this, activity))
+                    {
+                        await RunPipelineAsync(context, bot.OnTurnAsync, cancellationToken).ConfigureAwait(false);
+
+                        response.StatusCode = 200;
```

Comment:
There seems to be an issue here. I get "StatusCode cannot be set because the response has already started" when testing with the TestBot you have in the other branch.

Basically you can't set the status code this late (this code seems a bit complicated too, it may be good to refactor it so it is easier to follow.

Steps to repro the issue:

In BotController.cs in the test app, wrap the post async call in a try/catch and add a breakpoint in the exception:

![image](https://user-images.githubusercontent.com/12264946/62332260-c8469200-b48c-11e9-9c85-ef34a4dc7bf9.png)

Run the bot using twilio and send a message, you will get the response from twilio but you will also get an exception in the catch:

![image](https://user-images.githubusercontent.com/12264946/62332321-0fcd1e00-b48d-11e9-91a7-8783eec8046f.png)

Reply:
We just pushed the fix for this bug.
We are also working on the refactor of the method. We'll let you know when it's ready.
Thanks!